[inductor] make `conv2d` tests pass

nkaretnikov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #91952

```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

nkaretnikov committed:
[inductor] make `conv2d` tests pass

```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

[ghstack-poisoned]

nkaretnikov commented:
@ezyang Thoughts on this one? Not sure if it's cool, so a draft for now.

How I ended up here:
- `conv2d` on the C++ side calls `convolution_symint`
- `convolution_symint` calls `meta_conv` on the Python side
- `meta_conv` fails when calling `to` (fixed in this PR)
- but `mkldnn_reorder_conv2d_weight` is now called with a `SymInt` param
- couldn't easily fix `mkldnn_reorder_conv2d_weight` on the C++ side/in the yaml file (codegen problem?) -- was getting a dynamic linker error -- maybe can figure it out later, but...
-  using `SymInt` in the C++ code doesn't seem to make sense anyway because the code uses `ideep` functions, which are not `SymInt`-aware, so I just added manual guards on the Python side.

ezyang reviewed (approved):
Fine. But don't you need to remove an xfail/skip?

ezyang commented:
I think there's a decent chance that you don't want to guard on input size for mkldnn though; you need to actually work out how to make it dynamic.

nkaretnikov commented:
> Fine. But don't you need to remove an xfail/skip?

Which one?

ezyang commented:
Whatever is supposed to test this change (or add a new test?)

jgong5 commented:
> I think there's a decent chance that you don't want to guard on input size for mkldnn though; you need to actually work out how to make it dynamic.

It would be ideal that mkldnn can support weight prepacking (e.g., `mkldnn_reorder_conv2d_weight`) given dynamic shape input with some logical constraints as hint but unfortunately, it can only work with concrete shape input right now. Sorry about that... :P 
So, the change of guard on static sounds ok to me. cc @XiaobingSuper

nkaretnikov commented:
@ezyang So, these only failed when you ran with `TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1`. IIUC, CI doesn't use these vars yet. Is it okay to just add the following to the existing conv2d tests in test_torchinductor?

```
@patch.object(config, "dynamic_shapes", True)
@patch.object(torch._dynamo.config, "dynamic_shapes", True)
@patch.object(functorch_config, "use_dynamic_shapes", True)
```

Note that no other tests in test_torchinductor do this except dynamic-shapes-specific ones.

ezyang commented:
Ah, you want to do the trick we do in test/dynamo/test_dynamic_shapes.py where we clone the test to run under dynamic shapes config (assuming inductor doesn't do it already).

What you suggested will prevent this test from running in static configuration ever

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented:
Pushed an update that runs the tests that used to fail with dynamic shapes, regardless of whether you use the dynamic shapes env vars or not. Wanted a minimal fix, but reusing existing test infra turned out to be easier.

ezyang commented on test/inductor/test_torchinductor.py:
```diff
@@ -5058,6 +5058,25 @@ def fn(x):
+            (torch._dynamo.config, "dynamic_shapes", True),
+            (functorch_config, "use_dynamic_shapes", True),
+        ),
+    )
```
It feels better to do the same pattern as test_dynamic_shapes.py in torchdynamo, because in the end state, we want ALL of the tests passing, and having to manually reenumerate every test here to enable it testing on dynamic shapes is bad (because people will certainly forget). You have to do a big bang "disable most tests as not working on dynamic shapes" to start with (write a script to generate this list for you), but it puts us in a better spot for forward looking dev. 

cc @ngimel @Chillee for other opinions

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented:
Pushed an update based on a different PR in the stack. These tests should now return "Unexpected success" since they are not yet in the xfail list, to make sure that everything is working: test_conv2d_binary, test_conv2d_packed, test_conv2d_unary. Will add them later.

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
It's possible this will conflict with Horace's recent conv PR

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen chunyuan-w @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

[ghstack-poisoned]

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

[ghstack-poisoned]

nkaretnikov commented:
@Chillee We've discussed this offline a bit. Okay to merge?

One issue with conv tests is that they are slow (even non-dyn ones). All other tests take 2 mins to run, conv tests take like 10 mins. Are we cool with that?

Chillee commented:
@nkaretnikov Perhaps add the "slow tests" annotation (don't remember it off the top of my head). Otherwise LGTM!

nkaretnikov committed:
Update on "[inductor] make `conv2d` tests pass"


```
TORCHDYNAMO_DYNAMIC_SHAPES=1 AOT_DYNAMIC_SHAPES=1 python -m pytest -v test/inductor/test_torchinductor.py -k test_conv2d
```

[ghstack-poisoned]

nkaretnikov commented:
For our future selves: I'm going to merge this as is, without the `@slowTest` decorators. Slow tests are disabled by default and it's hard to understand when they actually run. So we could end up in a situation with no conv tests. Let's get back to this if someone complans.

```
pytest test/inductor/test_torchinductor.py -k conv -v --durations=0

these take more than 60s:

390.82s call     test/inductor/test_torchinductor.py::CpuTests::test_conv2d_unary_dynamic_shapes_cpu
303.10s call     test/inductor/test_torchinductor.py::CpuTests::test_conv2d_binary_dynamic_shapes_cpu
67.76s call     test/inductor/test_torchinductor.py::CpuTests::test_conv2d_unary_cpu
66.15s call     test/inductor/test_torchinductor.py::CpuTests::test_conv2d_binary_cpu
```

nkaretnikov commented:
@pytorchbot merge
